File: Regression/MultipleLinearRegression/multiple_linear_regression.py
# Multiple Linear Regression

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 4].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])

# To prevent machine learning model to think that Newyork > California > Florida use dummy variable.
# So we are gonna have three columns. First cloumn corresponds to e.g. France, second to Germany and third to Spain
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding Dummy Variable Trap. Removed first column. However python library 
# does that so we don't need to do it.
X = X[:, 1:]

# Splitting the dataset into Training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Feature Scaling
''' from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
# sc_X is already fit in training set so no need to fit for test set
X_test = sc_X.transform(X_test)'''

# Fitting Multiple Linear Regression to the Training Set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

# Predicting the Test set results
Y_pred = regressor.predict(X_test)

# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
# sm library doesn't account b0. So we need to add a column of 1s i.e. x0=1
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis=1)

File: Regression/MultipleLinearRegression/multiple_linear_regression.R
# Multiple Linear Regression

# Data Preprocessing

# Importing dataset
dataset = read.csv('50_Startups.csv')
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set =  subset(dataset, split == TRUE)
test_set =  subset(dataset, split == FALSE)
# Feature scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

# Fitting Multiple Linear Regression to Training set
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State)
# Same as above
regressor = lm(formula = Profit ~ .,
               data = training_set)

# By looking at the summary(regressor) only siginificant variable is R.D.Spend so we can reduce
# the above equation into simple linear regression. i.e. regressor = lm(formula = Profit ~ R.D.Spend, = training_set)

# Predicting the test set results.
Y_pred = predict(regressor, newdata = test_set)

# Check Y_pred and compare Y_pred with only significant variable. It'll almost be the same.
# regressor1 = lm(formula = Profit ~ R.D.Spend,
#               data = training_set)
# Y_pred = predict(regressor1, newdata = test_set)

# Building optimal model using Backward Elimination.
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = training_set)
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = training_set)
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = training_set)
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend,
               data = training_set)
summary(regressor)


